codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java

```diff
@@ -307,11 +366,133 @@ private boolean doRead0(Object msg, RecvByteBufAllocator.Handle allocHandle) {
         @Override
         public void connect(final SocketAddress remoteAddress,
                 SocketAddress localAddress, final ChannelPromise promise) {
             promise.setFailure(new UnsupportedOperationException());
         }
     }
+
+    /**
+     * Returns the flow-control size for DATA frames, and 0 for all other frames.
+     */
+    private static final class FlowControlledFrameSizeEstimator implements MessageSizeEstimator {
+
+        private static final FlowControlledFrameSizeEstimator INSTANCE = new FlowControlledFrameSizeEstimator();
+
+        private static final class EstimatorHandle implements MessageSizeEstimator.Handle {
+
+            private static final EstimatorHandle INSTANCE = new EstimatorHandle();
+
+            @Override
+            public int size(Object msg) {
+                if (msg instanceof Http2DataFrame) {
+                    return dataFrameFlowControlBytes((Http2DataFrame) msg);
+                }
+                return 0;
+            }
+        }
+
+        @Override
+        public Handle newHandle() {
+            return EstimatorHandle.INSTANCE;
+        }
+    }
+
+    /**
+     * {@link ChannelConfig} so that the high and low writebuffer watermarks can reflect the outbound flow control
+     * window, without having to create a new {@link WriteBufferWaterMark} object whenever the flow control window
+     * changes.
+     */
+    private final class Http2StreamChannelConfig extends DefaultChannelConfig {
+
+        // TODO(buchgr): Overwrite the RecvByteBufAllocator. We only need it to implement max messages per read.
+        Http2StreamChannelConfig(Channel channel) {
+            super(channel);
+        }
+
+        @Override
+        @Deprecated
+        public int getWriteBufferHighWaterMark() {
+            int window = (int) min(Integer.MAX_VALUE, outboundFlowControlWindow);
+            return max(0, window);
+        }
+
+        @Override
+        @Deprecated
+        public int getWriteBufferLowWaterMark() {
+            return getWriteBufferHighWaterMark();
+        }
+
+        @Override
+        public MessageSizeEstimator getMessageSizeEstimator() {
+            return FlowControlledFrameSizeEstimator.INSTANCE;
+        }
+
+        // TODO(buchgr): Throwing exceptions is not ideal. Maybe NO-OP and log a warning?
+        @Override
+        public WriteBufferWaterMark getWriteBufferWaterMark() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public ChannelConfig setMessageSizeEstimator(MessageSizeEstimator estimator) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        @Deprecated
+        public ChannelConfig setWriteBufferHighWaterMark(int writeBufferHighWaterMark) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        @Deprecated
+        public ChannelConfig setWriteBufferLowWaterMark(int writeBufferLowWaterMark) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the missing @deprecated Javadoc tag. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)